### DISCUSSION TITLE
Control SSerialRecieved MQTT payload with equipments that send data automatically on their own with any trigger/command sent by you.

### DISCUSSION THREAD

**Discussion by ortegafernando** [arendst/Tasmota#16537] [Q&A]
Hi, I am thinking about connecting a tasmota to a Victron MPPT (75/15).
Victron equipments usually send data over a serial port once a second (ve.direct protocol is used in low power mptt and inverters)

I have several options:
1. Write my own driver (I am not an specialist in tasmota code) that works with software serial and decode the message in tasmota and send data over mqtt. It will be the best option, of course, but we will need to compile our own version of bin file everytime tasmota has commits or a new release.
2. Use Software Serial Bridge functionally and recieve every second a message (SSerialRecieved) in my mqtt broker, and with NodeRed decode it in my server. Tasmota will be very easy to configure and update, but.... once a second is a lot of messages.
3. Deactivate the SSerialRecieved MQTT message with a tasmota option, and configure these rules: one for count SSerialRecievedData recieved, and every ... let's configure ....10 times the first rule will fire a publish command with the last serial data recieved. This way I will recieve a mqtt message to decode in my nodered once every 10 seconds, or once every .... what I need. It will be easy to configure and update, as this tasmota will have only the default firmware and two or three rules that are easy to configure.
4. Or is there any option to recive the SSerialRecieved event in rules without getting the same mqtt payload? It is basically the option 3 that I am asking for, but may be any configuration in tasmota diferent than serial bridge that could behaviour like I want.

What do you think about the option 3?
How could I make a PR with a new option to do this? May be it is not easy.

We have here the call that sends SSerialRecieved payload.
https://github.com/arendst/Tasmota/blob/2eefdaa0f40bb3f6b4c73ddb10ea31740eb163e0/tasmota/tasmota_xdrv_driver/xdrv_08_serial_bridge.ino#L170

This call is
https://github.com/arendst/Tasmota/blob/dc217655a090caa6658235fbb50f745133f7e19c/tasmota/tasmota_xdrv_driver/xdrv_02_9_mqtt.ino#L785

```
void MqttPublishPrefixTopicRulesProcess_P(uint32_t prefix, const char* subtopic, bool retained) {
  // Publish <prefix>/<device>/<RESULT or <subtopic>> default ResponseData string with optional retained
  //   then process rules
  MqttPublishPrefixTopic_P(prefix, subtopic, retained);
  XdrvRulesProcess(0);
}
```

It could be done only with this change in xdrv_08_serial_bridge.ino?:
```
if ( !Settings ->flagX.mqtt_no_sserialrecieved) {   //If I don't want the SSerialReceived MQTT message...don't send it
     MqttPublishPrefixTopicRulesProcess_P(RESULT_OR_TELE, PSTR(D_JSON_SSERIALRECEIVED));
}
```
Will I still recieved the SSerialREceived event that can be processed in rules?
Could any of the experts of tasmota add this option in the default code?

Thanks a lot for your help.

NOTE: If anyone is interest, I could add in this post, the nodered source code to decode the message that it recieves. I haven't yet started, but it will be very easy.

**Comment by sfromis** [comment]
What I'd prefer for controlling local processing of serial data would be to use ESP32 and a small Berry script to listen to the serial data, and do whatever filtering and publishing, possibly including other transformation at the same time. This is quite a bit easier than creating a lower level driver embedded into Tasmota.

**Reply by ortegafernando** [reply]
> What I'd prefer for controlling local processing of serial data would be to use ESP32 and a small Berry script to listen to the serial data, and do whatever filtering and publishing, possibly including other transformation at the same time. This is quite a bit easier than creating a lower level driver embedded into Tasmota.

Thanks sfromis, but I have several esp8266 in my house and I want to "use" them before buying anything more. 

Even I prefer to "code" in nodered "cloud" vs Berry script in order to have a "simplier tasmota device", and having a "cloud decode" service with which I could update/change/debug/... the code easier than code "inside" tasmota. (of course there will be some cases that it is better to do in berry, but a simple bridge between serial port and "wifi/mqtt" I prefer to keep in simply in tasmota.

With an esp8266 what do you think I could do?

**Reply by barbudor** [reply]
I don't see 1 msg per second as too much. Neither for Tasmota to send nor for node-red to manage. I would forsee more problems with rules execution. 
Receive and publish is IMHO much easier than Receive and Process

A config to send only 1 msg out of N doesn't make much sense to me

I subscribe to @sfromis suggestion with Berry. With the existance of ESP32c3 modules in the Esp12 fo factor, brain transplant may be an option

**Reply by ortegafernando** [reply]
> I don't see 1 msg per second as too much. Neither for Tasmota to send nor for node-red to manage. I would forsee more problems with rules execution. Receive and publish is IMHO much easier than Receive and Process
> 
> A config to send only 1 msg out of N doesn't make much sense to me
> 
> I subscribe to @sfromis suggestion with Berry. With the existance of ESP32c3 modules in the Esp12 fo factor, brain transplant may be an option

I will receive and publish only. I will not process anything in tasmota. Rules are only to "count" how many "receive"'s and to publish once every X times.

I understand you about Berry, but, please, understand me. I have several esp8266 arround and I also prefer to "code/process" in my nodered server than in every device specifically. These two reasons make me to try ESP8266 over Berry and ESP32.

Do you think it is very hard to implement a new setting like I am describing in option 3? I will only waste 1 bit of memory and a few bytes of code in tasmota default bin file

**Reply by sfromis** [reply]
What you want is to process it in Tasmota, instead of "receive and publish *only*". That counting is very much local processing.

Anyway, if it is only one message a second, that should be no problem to Tasmota, your network, your broker, nor a reasonably done Node-Red setup. Are you actually experiencing a problem, or are you just thinking "I really do not need this much"?

Note that nothing stops you from doing local changes to the Tasmota code, to adapt to your very specific setup and ideas.

**Reply by ortegafernando** [reply]
> What you want is to process it in Tasmota, instead of "receive and publish _only_". That counting is very much local processing.
> 
> Anyway, if it is only one message a second, that should be no problem to Tasmota, your network, your broker, nor a reasonably done Node-Red setup. Are you actually experiencing a problem, or are you just thinking "I really do not need this much"?
> 
> Note that nothing stops you from doing local changes to the Tasmota code, to adapt to your very specific setup and ideas.

Ja ja, you are right. I am thinking that I really dont need this much (also, my sever is not local, it is in cloud). I think it is better not to saturate if you don't need. 

Of course about change tasmota code (that is option 1 in my first post) but I have some disadvantages. 

I supposed that an event rule only to "add 1" to a counter is not much local processing.

Is that rule (count events fired) so much processing time?

Do you think that if I make a PR with a new setting it will be rejected or accepted? 

Thanks.

**Reply by sfromis** [reply]
The processing time for counting should indeed be no issue, when your plan is to do local processing.

I don't know if such a PR would be accepted, it could also be an optional build time feature not affecting other builds.

**Reply by ortegafernando** [reply]
> The processing time for counting should indeed be no issue, when your plan is to do local processing.
> 
> I don't know if such a PR would be accepted, it could also be an optional build time feature not affecting other builds.

Sorry, I dont understand you first paragraph.

So counting is not a problem, but "when your plan is to do local processing"... not, my plan is count locally (esp8266) and process in cloud (nodered/mosquitto server).

I will wait more comments to see if they will find the PR usefull. May be there are more products, sensors, ... that send data over serial port without any interaction/trigger sent by tasmota.

🙏

I think that will be only a bit and very few bytes of code (only a "if")

**Reply by sfromis** [reply]
I was merely reiterating that the counting is very much you wanting to have "local processing", and agreeing with you that the processing time should be no issue.

**Reply by ortegafernando** [reply]
PR created and wait for admin moderation.

https://github.com/arendst/Tasmota/pull/16612
